subsys/net/lib/mqtt/mqtt.c

```diff
@@ -726,70 +727,12 @@ int mqtt_publisher_parser(struct mqtt_ctx *ctx, struct net_buf *rx)
 	case MQTT_PINGRESP:
 		rc = mqtt_rx_pingresp(ctx, data);
 		break;
-	default:
-		rc = -EINVAL;
-		break;
-	}
-
-exit_parser:
-	if (rc != 0 && ctx->malformed) {
-		ctx->malformed(ctx, pkt_type);
-	}
-
-	net_nbuf_unref(data);
-
-	return rc;
-}
-
-
-/**
- * Calls the appropriate rx routine for the MQTT message contained in rx
- *
- * @details On error, this routine will execute the 'ctx->malformed' callback
- * (if defined)
- *
- * @param ctx MQTT context
- * @param rx RX buffer
- *
- * @retval 0 on success
- * @retval -EINVAL if an unknown message is received
- * @retval -ENOMEM if no data buffer is available
- * @retval mqtt_rx_publish, mqtt_rx_pubrel, mqtt_rx_pubrel and mqtt_rx_suback
- *         return codes
- */
-static
-int mqtt_subscriber_parser(struct mqtt_ctx *ctx, struct net_buf *rx)
-{
-	uint16_t pkt_type = MQTT_INVALID;
-	struct net_buf *data = NULL;
-	int rc = 0;
-
-	data = mqtt_linearize_buffer(ctx, rx, MQTT_PUBLISHER_MIN_MSG_SIZE);
-	if (!data) {
-		rc = -EINVAL;
-		goto exit_parser;
-	}
-
-	pkt_type = MQTT_PACKET_TYPE(data->data[0]);
-
-	switch (pkt_type) {
-	case MQTT_CONNACK:
-		if (!ctx->connected) {
-			rc = mqtt_rx_connack(ctx, data, ctx->clean_session);
-		} else {
-			rc = -EINVAL;
-		}
-
-		break;
 	case MQTT_PUBLISH:
 		rc = mqtt_rx_publish(ctx, data);
 		break;
 	case MQTT_PUBREL:
 		rc = mqtt_rx_pubrel(ctx, data);
 		break;
-	case MQTT_PINGRESP:
-		rc = mqtt_rx_pubrel(ctx, data);
-		break;
 	case MQTT_SUBACK:
```

Comment:
Combining these into one function seems like it turns every app into an MQTT_APP_PUBLISHER_SUBSCRIBER. A publish-only app wouldn't expect to ever receive SUBACK, so I'd expect this to error out here in that case, and similarly in other situations (receiving PUBACK or PUBCOMP in a subscribe-only app, etc.)

Am I missing something?